### ISSUE TITLE
Packaging broken

### ISSUE THREAD

**Issue by asymmetric** [muesli/beehive-vendor#2]
Hi,

First of all, thanks for the nice project!

I'd like to create a Bee, but the packaging process is currently pretty broken.

What's the point of using a package manager, and committing `vendor` to git, when the strategy is

* Have dependencies in a separate repo, loaded as a submodule
* Not to commit the `Gopkg.lock`
* Delete and replace the whole `vendor/` dir

Are there any reasons I'm missing for this non-idiomatic approach?

To the point of being broken: the `github.com/jteeuwen/go-pkg-rss` is not on GitHub anymore. This means that simply running `dep init` fails, as that dependency cannot be found.

I've seen you've [forked](https://github.com/muesli/go-pkg-rss) the repo, so I guess the next thing to do is change imports in bee.

But before I open a PR for that, I'd like to propose that Bee moves to a more standard vendoring workflow:

* Committing the `Gopkg.toml` and `Gopkg.lock` files
* Just using `dep ensure` on the main repo
* Deleting this repo

**Comment by asymmetric** [comment]
BTW, I've started working on this on my [own branch](https://github.com/muesli/beehive/compare/master...asymmetric:asymmetric/dependencies).

I haven't created a PR because I removed the RSS dependency in my fork, and I'm not sure that's the way to go.

The alternative would be to change every reference to `go-pkg-rss` and `go-pkg-xmlx` in bee.

**Comment by muesli** [comment]
At the time when we started using `dep` it was actually recommended _not_ to commit the `Gopkg.toml` and `Gopkg.lock` files (even though that seems to have changed by now). Also, `dep` requires a newer Go version than Beehive, so we didn't want to make it a build-time dependency (yet).

> Have dependencies in a separate repo, loaded as a submodule

I don't see an issue with using a submodule for vendoring, esp. in the case of Beehive which has tons of dependencies. I really dislike the idea of polluting the main repo with all the vendored sources.

> Delete and replace the whole vendor/ dir

I don't think we ever delete or replace the `vendor` dir. What are you referring to?

You're correct that we need to fix the `jteeuwen` mess, but since it's still vendored I didn't consider master broken.

I'd suggest committing `Gopkg.toml` and `Gopkg.lock`, dropping this repository and the submodule, and adding a `dep ensure` step to the build process. I'm still opposed to adding all the vendored sources to the main repo, though.

**Comment by asymmetric** [comment]
>  dep requires a newer Go version than Beehive, so we didn't want to make it a build-time dependency

AFAIK, `dep` only requires 1.9 if you plan on building it, not using it.

> I don't think we ever delete or replace the vendor dir. What are you referring to?

I'm referring to `beehive-vendor`'s [README](https://github.com/muesli/beehive-vendor/blob/master/README.md#updating-beehive-dependencies):

> 3. Delete the vendor folder

> since it's still vendored I didn't consider master broken.

If I follow the steps in the README, delete `vendor/` and run `dep init`, I cannot successfully fetch dependencies, since that package has been removed.

> I'd suggest committing Gopkg.toml and Gopkg.lock, dropping this repository and the submodule, and adding a dep ensure step to the build process. I'm still opposed to adding all the vendored sources to the main repo, though.

I agree with you that it's unpleasant, but having been bitten by this in the past, I don't trust `dep`'s dependency resolution algorithm to guarantee reproducible builds. For that reason, I now commit `vendor/` with the rest of the repo, while I wait for `vgo` to be ready :)

**Comment by muesli** [comment]
Oh, I see! Sorry for the misunderstanding. I thought you still had issues building Beehive itself, which should work fine with `go get` / `git clone` checking out the vendor submodule for you automatically.

Updating the `beehive-vendor` repository currently is a bit problematic indeed, but for what it's worth I've fixed the `jteeuwen` imports in Beehive itself.

Can't wait for vgo either ;-)

**Comment by asymmetric** [comment]
Are you still open to switching to a `Gopkg.toml` setup, either with or without committing `vendor/` to the main repo, and removing this submodule?

I could open a PR for that.

**Comment by muesli** [comment]
> Are you still open to switching to a Gopkg.toml setup, either with or without committing vendor/ to the main repo, and removing this submodule?

Absolutely, as mentioned I think we could now commit Gopkg.toml to the main repo and drop the vendor repository. I still would prefer not committing `vendor` to the main repo, but we can change the build process (Makefile) and support dep directly instead.

That would mean a `go get` would fetch all dependencies and build a beehive from git master (including all its dependencies), while a `make` would fetch the dependencies with `dep` first and then builds a more reproducible build.

> I could open a PR for that.

Awesome! Looking forward to it!

**Comment by asymmetric** [comment]
Not including `vendor/` will make builds slower though, since every time dependencies will have to be fetched. Also, as mentioned previously, I'm not convinced builds are always reproducible with `dep`, meaning that bug reports could be hard to debug.

Also, `dep` will have to be downloaded within Travis, so the `.travis.yml` file will have to be kept in sync when newer `dep` versions are releasde.
